### DISCUSSION TITLE
paraview-visualizer losing connection

### DISCUSSION THREAD

**Discussion by earlsioson** [Kitware/trame#66] [Q&A]
Hello,

When running paraview-visualizer in a VM we periodically receive the following error
```
ERROR:asyncio:Task exception was never retrieved
future: <Task finished name='Task-51' coro=<WebSocketWriter.ping() done, defined at /opt/paraview/lib/python3.9/site-packages/aiohttp/http_websocket.py:669> exception=ConnectionResetError('Cannot write to closing transport')>
Traceback (most recent call last):
  File "/opt/paraview/lib/python3.9/site-packages/aiohttp/http_websocket.py", line 673, in ping
    await self._send_frame(message, WSMsgType.PING)
  File "/opt/paraview/lib/python3.9/site-packages/aiohttp/http_websocket.py", line 650, in _send_frame
    self._write(header + message)
  File "/opt/paraview/lib/python3.9/site-packages/aiohttp/http_websocket.py", line 660, in _write
    raise ConnectionResetError("Cannot write to closing transport")
ConnectionResetError: Cannot write to closing transport
```

Any ideas on what's happening?

**Comment by jourdain** [comment]
How do you run `paraview-visualizer` in your VM? Are you using the docker image or do you start it by hand?

Also I'm hoping to get back to `paraview-visualizer` sometime next week to push it further and maybe if I can reproduce your issue, I will be able to track it down to fix it...

**Reply by earlsioson** [reply]
@jourdain, I'm using the run_server.sh script in the docker/scripts directory.

**Reply by jourdain** [reply]
Can you provide any other context? Is it once the client disconnect? Is it while you are using the application? Is it because you have a filter that takes more than 30s to execute?

**Reply by earlsioson** [reply]
@jourdain,
It happens after the client is connected. I'm able to navigate to a data directory and select a can.ex2 example. Then after a few minutes I get this screen...
<img width="763" alt="image" src="https://user-images.githubusercontent.com/13466524/167706400-5c6a7f53-367b-4871-b2c0-8aec814e2a5c.png">
The error in the server/logs/launcher log is...
```
wslink: Starting factory
INFO:trame_app.app.engine:>>> ENGINE: Server protocols initialized / Client not connected yet
ERROR:asyncio:Task exception was never retrieved
future: <Task finished name='Task-6' coro=<WebSocketWriter.ping() done, defined at /deploy/server/venv/lib/python3.9/site-packages/aiohttp/http_websocket.py:672> exception=ConnectionResetError('Cannot write to closing transport')>
Traceback (most recent call last):
  File "/deploy/server/venv/lib/python3.9/site-packages/aiohttp/http_websocket.py", line 676, in ping
    await self._send_frame(message, WSMsgType.PING)
  File "/deploy/server/venv/lib/python3.9/site-packages/aiohttp/http_websocket.py", line 653, in _send_frame
    self._write(header + message)
  File "/deploy/server/venv/lib/python3.9/site-packages/aiohttp/http_websocket.py", line 663, in _write
    raise ConnectionResetError("Cannot write to closing transport")
ConnectionResetError: Cannot write to closing transport
```

**Reply by jourdain** [reply]
Interesting... That helps to narrow down the possible issue. Are you connecting directly to your docker image (localhost) or are you going through some proxy (nginx/apache/else)?

**Reply by earlsioson** [reply]
We are going through a Google Cloud Global Load Balancer.

**Reply by jourdain** [reply]
The sessionURL returned by the given docker image that the load balancer picked should be specialized and the ws should not go through the load balancer.

**Reply by earlsioson** [reply]
@jourdain,

Is that a requirement that the trame docker image is not to be used behind a Load Balancer or a recommendation?

The load balancer we are using has native support for websocket protocol, https://cloud.google.com/load-balancing/docs/https#websocket_support and we are using other websocket enabled apps via GCP load balancers and the connections are maintained.

Is there a configuration setting in Apache or the launcher we can try to prevent the connection from being disconnected?

**Reply by jourdain** [reply]
It is totally fine to use those docker images behind a load balancer but ideally you want the ws to reach back directly to the selected host and skip an extra hop. Doing so would require some extra config/edit inside the launcher config (and maybe GCP). 

If you want to keep using the load balancer to proxy the ws, you will need to reduce the time for the heartbeat to make sure you are not getting disconnected. And that also mean that your server can not be busy more than the heartbeat time that you choose. Which will depend on what you are doing with ParaView but I would think it could happen with somewhat big processing. 

For now you can configure the heartbeat using that [environment variable](https://github.com/Kitware/wslink/blob/master/python/src/wslink/backends/aiohttp/__init__.py#L23).

HTH

**Reply by earlsioson** [reply]
@jourdain Thank you for that info. That does help.

**Reply by jourdain** [reply]
Otherwise the line to edit would be the equivalent of [that one](https://github.com/Kitware/trame/blob/master/docker/config/default-launcher.json#L8) but within the generated `/server/` directory for the docker to use. And you will have to pass env property when running docker similar to what is done [here](http://kitware.github.io/paraviewweb/docs/docker.html#Run-ParaView-Docker-image) with **SERVER_NAME** and **PROTOCOL** while using that [sessionURL](https://github.com/Kitware/paraviewweb-demo/blob/master/pvw/launcher/config.json#L10) (using the SESSION_URL_ROOT prefix).
